relentless/core/variable.py

```diff
@@ -0,0 +1,311 @@
+        except AttributeError:
+            self._low = low
+
+        try:
+            self._value, self._state = self.clamp(self._value)
+        except AttributeError:
+            pass
+
+    @property
+    def high(self):
+        """float: The high bound of the variable"""
+        return self._high
+
+    @high.setter
+    def high(self, high):
+        if high is not None and not isinstance(high, (float,int)):
+            raise ValueError('High bound must be a float or int')
+        try:
+            if high is None or self._high is None:
+                self._high = high
+            elif high > self._low:
+                self._high = high
+            else:
+                raise ValueError('The high bound must be greater than the low bound')
+        except AttributeError:
+            self._high = high
+
+        try:
+            self._value, self._state = self.clamp(self._value)
+        except AttributeError:
+            pass
+
+    @property
+    def state(self):
+        """:py:class:`Variable.State`: The state of the variable"""
+        return self._state
+
+    def isfree(self):
+        """Confirms if the variable is unconstrained or within the bounds.
+
+        Returns
+        -------
+        bool
+            True if the variable is unconstrained or within the bounds, False otherwise.
+
+        """
+        return self.state is DesignVariable.State.FREE
+
+    def atlow(self):
+        """Confirms if the variable is at the lower bound.
+
+        Returns
+        -------
+        bool
+            True if the variable is at the lower bound, False otherwise.
+
+        """
+        return self.state is DesignVariable.State.LOW
+
+    def athigh(self):
+        """Confirms if the variable is at the upper bound.
+
+        Returns
+        -------
+        bool
+            True if the variable is at the upper bound, False otherwise.
+
+        """
+        return self.state is DesignVariable.State.HIGH
+
+class DependentVariable(Variable):
+    """Abstract base class for a variable that depends on other variables.
+
+    A dependent variable is composed from other variables. In addition to the
+    value property, it must also supply a derivative method with respect to
+    any of its dependent variables.
+
+    The number of dependencies of the variable are locked at construction, but
+    their values can be modified.
+
+    Parameters
+    ----------
+    vardicts : dict
+        Attributes for the variable on which the DependentVariable depends (positional argument).
+    kwvars : kwargs
+        Attributes for the variable on which the DependentVariable depends (keyword arguments).
+
+    """
+    def __init__(self, *vardicts, **kwvars):
+        attrs = {}
+        for d in vardicts:
+            attrs.update(d)
+        attrs.update(**kwvars)
+
+        for k,v in attrs.items():
+            super().__setattr__(k,self._assert_variable(v))
+        self._depends = tuple(kwvars.keys())
```

Comment:
Was just looking for some code and noticed something. Does this need to be `attr.keys()`? What happens if no `kwvars` are given? It would be good to add some tests with a mock dependent variable (like you did for the pair potential) that is constructed using different combinations of these args.